code/datums/components/anti_magic.dm

```diff
@@ -1,54 +1,120 @@
-		allowed_slots = _allowed_slots
-	if(!isnull(_charges))
-		charges = _charges
-	blocks_self = _blocks_self
-	reaction = _reaction
-	expire = _expire
+	src.antimagic_flags = antimagic_flags
+	src.charges = charges
+	src.inventory_flags = inventory_flags 
+	src.drain_antimagic = drain_antimagic
+	src.expiration = expiration
 
 /datum/component/anti_magic/proc/on_equip(datum/source, mob/equipper, slot)
 	SIGNAL_HANDLER
 
-	if(!(allowed_slots & slot)) //Check that the slot is valid for antimagic
+	if(!(inventory_flags & slot)) //Check that the slot is valid for antimagic
 		UnregisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC)
+		UnregisterSignal(equipper, COMSIG_MOB_RESTRICT_MAGIC)
+		equipper.update_action_buttons()
 		return
-	RegisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC, .proc/protect, TRUE)
+	RegisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC, .proc/block_receiving_magic)
+	RegisterSignal(equipper, COMSIG_MOB_RESTRICT_MAGIC, .proc/restrict_casting_magic)
+	equipper.update_action_buttons()
+
+	// Check to see if we have any spells that are blocked due to antimagic
+	for(var/datum/action/spell_action/spell/magic_action in equipper.actions)
+		var/obj/effect/proc_holder/spell/magic_spell = magic_action.target
+		if(antimagic_flags & magic_spell.antimagic_flags)
+			to_chat(equipper, span_warning("[parent] is interfering with your ability to cast magic!"))
+			break
+
 
 /datum/component/anti_magic/proc/on_drop(datum/source, mob/user)
 	SIGNAL_HANDLER
 
 	UnregisterSignal(user, COMSIG_MOB_RECEIVE_MAGIC)
+	UnregisterSignal(user, COMSIG_MOB_RESTRICT_MAGIC)
+	user.update_action_buttons()
+
+/datum/component/anti_magic/proc/block_receiving_magic(datum/source, mob/user, casted_magic_flags, charge_cost)
+	SIGNAL_HANDLER
+
+	// disclaimer - All anti_magic sources will be drained a charge_cost
+	if(casted_magic_flags & antimagic_flags)
+		// im a programmer not shakesphere to the future grammar nazis that come after me for this
+		var/visible_subject = ismob(parent) ? "[user.p_they()]" : "[parent]"
+		var/self_subject = ismob(parent) ? "you" : "[parent]"
+		if(casted_magic_flags & antimagic_flags & MAGIC_RESISTANCE)
+			user.visible_message(span_warning("[user] pulses red as [visible_subject] absorbs magic energy!"), \
+			span_userdanger("An intense magical aura pulses around [self_subject] as it dissipates into the air!"))
+		else if(casted_magic_flags & antimagic_flags & MAGIC_RESISTANCE_HOLY)
+			user.visible_message(span_warning("[user] starts to glow as [visible_subject] emits a halo of light!"), \
+			span_userdanger("A feeling of warmth washes over [self_subject] as rays of light surround your body and protect you!"))
+		else if(casted_magic_flags & antimagic_flags & MAGIC_RESISTANCE_MIND)
+			user.visible_message(span_warning("[user] forehead shines as [visible_subject] repulses magic from their mind!"), \
+			span_userdanger("A feeling of cold splashes on [self_subject] as your forehead reflects magic targeting your mind!"))
```

Comment:
All of these visible messages should be put in `can_block_magic`, where the effects are done. 

Currently, the mob level displays an effect (red shield, halo) while the visible message aspect is handled by this component. They should both be in the same place.

Reply:
I already considered this and decided it wasn't practical.  

The reason we put the visible messages in the antimagic component is so we have access to the `parent` object.  `can_block_magic` does not have access to that.  Also a person can have several different antimagic items on that block the magic and we want to display each one triggering when magic is blocked. This is why the visual/sound effect is only triggered once in `can_block_magic`.

Reply:
You could send `parent` back from the signal the way the old signal did - `protections_list`. Signal sends a list in place and gets back a list of all atoms that contributed to blocking the spell. 

I really don't like that these visible messages are decoupled from the effects - If someone overrides the proc to not show effects, these will still show, and if someone adds the ability to block magic from another way, these messages will not show.

Reply:
Applied with https://github.com/tgstation/tgstation/pull/64124/commits/ac2a463dbe2457e5a0412ac8fc7025acce5cb3d7